src/libraries/System.Private.CoreLib/src/System/Text/TranscodingStream.cs

```diff
@@ -0,0 +1,603 @@
+         * if this instance has been disposed. CanSeek is always false.
+         */
+
+        public override bool CanRead => _innerStream?.CanRead ?? false;
+
+        public override bool CanSeek => false;
+
+        public override bool CanWrite => _innerStream?.CanWrite ?? false;
+
+        public override long Length => throw new NotSupportedException(SR.NotSupported_UnseekableStream);
+
+        public override long Position
+        {
+            get => throw new NotSupportedException(SR.NotSupported_UnseekableStream);
+            set => throw new NotSupportedException(SR.NotSupported_UnseekableStream);
+        }
+
+        public override IAsyncResult BeginRead(byte[] buffer, int offset, int count, AsyncCallback? callback, object? state)
+            => TaskToApm.Begin(ReadAsync(buffer, offset, count, CancellationToken.None), callback, state);
+
+        public override IAsyncResult BeginWrite(byte[] buffer, int offset, int count, AsyncCallback? callback, object? state)
+            => TaskToApm.Begin(WriteAsync(buffer, offset, count, CancellationToken.None), callback, state);
+
+        protected override void Dispose(bool disposing)
+        {
+            Debug.Assert(disposing, "This type isn't finalizable.");
+
+            if (_innerStream is null)
+            {
+                return; // dispose called multiple times, ignore
+            }
+
+            // First, flush any pending data to the inner stream.
+
+            ArraySegment<byte> pendingData = FinalFlushWriteBuffers();
+            if (pendingData.Count != 0)
+            {
+                _innerStream.Write(pendingData);
+            }
+
+            // Mark our object as disposed
+
+            Stream innerStream = _innerStream;
+            _innerStream = null!;
+
+            // And dispose the inner stream if needed
+
+            if (!_leaveOpen)
+            {
+                innerStream.Dispose();
+            }
+        }
+
+        public override ValueTask DisposeAsync()
+        {
+            if (_innerStream is null)
+            {
+                return default; // dispose called multiple times, ignore
+            }
+
+            // First, get any pending data destined for the inner stream.
+
+            ArraySegment<byte> pendingData = FinalFlushWriteBuffers();
+
+            if (pendingData.Count == 0)
+            {
+                // Fast path: just dispose of the object graph.
+                // No need to write anything to the stream first.
+
+                Stream innerStream = _innerStream;
+                _innerStream = null!;
+
+                return (_leaveOpen)
+                    ? default /* no work to do */
+                    : innerStream.DisposeAsync();
+            }
+
+            // Slower path; need to perform an async write followed by an async dispose.
+
+            return DisposeAsyncCore(pendingData);
+            async ValueTask DisposeAsyncCore(ArraySegment<byte> pendingData)
```

Comment:
Local functions aren't callable outside the method which declares them. Similar to how the backing fields of auto-properties aren't accessible outside the property which declares them. It helps clearly denote the local method as "should never be used by anybody except for its containing parent."

Most of the time it's because the parent has a fast-path which should be inlined into the caller, or it's because the parent needs to perform argument checks before the async state machine kicks in.